src/phpDocumentor/Dsn.php

```diff
@@ -0,0 +1,308 @@
+<?php
+/**
+ * This file is part of phpDocumentor.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @copyright 2010-2015 Mike van Riel<mike@phpdoc.org>
+ * @license   http://www.opensource.org/licenses/mit-license.php MIT
+ * @link      http://phpdoc.org
+ */
+
+namespace phpDocumentor;
+
+/**
+ * Value Object for DSN.
+ */
+
+final class Dsn
+{
+    /** @var string */
+    private $scheme;
+
+    /** @var string */
+    private $host;
+
+    /** @var int */
+    private $port;
+
+    /** @var string */
+    private $user;
+
+    /** @var string */
+    private $password;
+
+    /** @var string */
+    private $path;
+
+    /** @var array  */
+    private $query = [];
+
+    /** @var array  */
+    private $parameters = [];
+
+    /** @var array  */
+    private $locationParts = [];
+
+    /** @var array  */
+    private $dsnParts = [];
+
+    /**
+     * @param $dsn
+     */
+    public function __construct($dsn)
+    {
+        $this->parse($dsn);
+    }
+
+    /**
+     * @return string
+     */
+    public function getScheme()
+    {
+        return $this->scheme;
+    }
+
+    /**
+     * @return string
+     */
+    public function getHost()
+    {
+        return $this->host;
+    }
+
+    /**
+     * @return int
+     */
+    public function getPort()
+    {
+        return $this->port;
+    }
+
+    /**
+     * @return string
+     */
+    public function getUsername()
+    {
+        return $this->user;
+    }
+
+    /**
+     * @return string
+     */
+    public function getPassword()
+    {
+        return $this->password;
+    }
+
+    /**
+     * @return string
+     */
+    public function getPath()
+    {
+        return new Path($this->path);
+    }
+
+    /**
+     * @return string
+     */
+    public function getQuery()
+    {
+        return $this->query;
+    }
+
+    /**
+     * @return string
+     */
+    public function getParameters()
+    {
+        return $this->parameters;
+    }
+
+    /**
+     * @param $dsn
+     */
+    private function parse($dsn)
+    {
+        $this->dsnParts = explode(';', $dsn);
+        $location = $this->dsnParts[0];
+        unset($this->dsnParts[0]);
+        $this->locationParts = parse_url($location);
```

Comment:
In your parsing methods there are a fair number of regular expressions to validate parts of your DSN. Regular expressions can generally be quite slow and in this case we can improve performance, and readability, with the following optimization:

Before parsing your parts you check if the scheme is set (thus after the parse_url call) and if so: prefix the $location with `file://`. The result of $location should then form a valid URL which you can check with http://php.net/manual/en/function.filter-var.php

After that you have validated the complete URL and you don't need to do validations of the separate components